btree/btree.c

```diff
@@ -2035,11 +2191,877 @@ static void generic_move(struct nd *src, struct nd *tgt,
+		2)last key of left node needs to be set as empty(if required)
+		*/
+
+		/* 1)update bop->bo_rec which needs to be added at parent */
+		l_slot.s_idx = node_count(l_slot.s_node);
+		l_slot.s_rec = temp_rec;
+		node_key(&l_slot);
+
+		m0_bufvec_copy(&bop->bo_rec.r_key.k_data,
+			       &l_slot.s_rec.r_key.k_data,
+			       m0_vec_count(&l_slot.s_rec.r_key.k_data.ov_vec));
+		temp_rec.r_val.ov_buf[0] =  &(curr_level->l_alloc->n_addr);
+		m0_bufvec_copy(&bop->bo_rec.r_val,
+				&temp_rec.r_val, 8);
+		//@TBD:replace 8 with vsize for internal node
+
+		/*
+		2)update last key of left node to null/empty(optional)
+		(for internal node, if it is strictly required to make last key
+		as null, perform this step) as we move items to right node, last
+		key won't be null
+		*/
+		/*modify_key(&slot_for_left_node, false, bop->bo_tx);
+		node_done(tgt, bop->bo_tx, true);
+		node_fix(tgt->s_node, bop->bo_tx);
+		curr_level->l_alloc->n_seq++;*/
+	} else {
+		/*
+		If we are at leaf node, update record (bop->rec) which needs to
+		get inserted at parent.
+		1)key of new record will be first key from right node and
+		2)value of new record will be pointer to l_alloc;
+		*/
+		r_slot.s_idx = 0;
+		r_slot.s_rec = temp_rec;
+		node_key(&r_slot);
+
+		m0_bufvec_copy(&bop->bo_rec.r_key.k_data,
+			       &r_slot.s_rec.r_key.k_data,
+			       m0_vec_count(&r_slot.s_rec.r_key.k_data.ov_vec));
+		temp_rec.r_val.ov_buf[0] =  &(curr_level->l_alloc->n_addr);
+		m0_bufvec_copy(&bop->bo_rec.r_val,
+				&temp_rec.r_val, 8);
+		//@TBD:replace 8 with vsize for internal node
+	}
+	return P_NEXTUP;
+}
+
+/**
+ * This function gets called when splitting is done at root node. This function
+ * is responsible to handle this scanario and ultimately root will point out to
+ * the two splitted node.
+ * @param bop it structure for btree operation which contains all required data
+ * @return int it return state which needs to get executed next
+ */
+static int m0_btree_put_root_split_handle(struct m0_btree_op *bop)
+{
+	struct td              *tree       = bop->bo_arbor->t_desc;
+	struct m0_btree_oimpl  *oi         = bop->bo_i;
+	struct level           *curr_level = &oi->i_level[oi->i_used];
+	m0_bcount_t             ksize;
+	void                   *p_key;
+	m0_bcount_t             vsize;
+	void                   *p_val;
+	struct m0_btree_rec     temp_rec;
+
+	temp_rec.r_key.k_data = M0_BUFVEC_INIT_BUF(&p_key, &ksize);
+	temp_rec.r_val        = M0_BUFVEC_INIT_BUF(&p_val, &vsize);
+
+	/*
+	When splitting is done at root node, tree height needs to get increased
+	by one. As, we do not want to change the pointer to the root node, we
+	will copy all contents from root to i_extra_node and make i_extra_node
+	as one of the child of existing root
+	1) First copy all contents from root node to extra_node
+	2) add new 2 records at root node:
+		i.for first record, key = bop->bo_rec.r_key,
+		  value = bop->bo_rec.r_val
+		ii.for first record, key = null, value = segaddr(i_extra_node)
```

Comment:
Consider scenario, when there is splitting of root, there will be two new nodes at root node(root, l_alloc), and we need to store the address of these splitted node into level above the root. But there will not be any nodes above the root node, so we use one more node(i_extra_node) to handle this case , where as we do not want to change the root, we use i_extra_node and copy the content from root to i_extra_node, now two nodes which needs to get added at root will be l_alloc, i_extra_node. So, as root will become an internal node it will have one key but two values, last key it can be null or anything(depends on developer) but for any internal node, we will not count the last key though it will be in memory